app/api/items/save_grade.go

```diff
@@ -0,0 +1,275 @@
+package items
+
+import (
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/formdata"
+	"github.com/France-ioi/AlgoreaBackend/app/payloads"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+	"github.com/France-ioi/AlgoreaBackend/app/token"
+)
+
+func (srv *Service) saveGrade(w http.ResponseWriter, r *http.Request) service.APIError {
+	requestData := SaveGradeRequest{store: srv.Store, publicKey: srv.TokenConfig.PublicKey}
+
+	var err error
+	if err = render.Bind(r, &requestData); err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+	if err = user.Load(); err == database.ErrUserNotFound {
+		return service.InsufficientAccessRightsError
+	}
+	service.MustNotBeError(err)
+
+	apiError := service.NoError
+	if apiError = checkSaveGradeTokenParams(user, &requestData); apiError != service.NoError {
+		return apiError
+	}
+
+	var validated, keyObtained bool
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		var hasAccess bool
+		var reason error
+		hasAccess, reason, err = store.Items().CheckSubmissionRights(requestData.TaskToken.Converted.LocalItemID, user)
+		service.MustNotBeError(err)
+
+		if !hasAccess {
+			apiError = service.ErrForbidden(reason)
+			return nil // commit! (CheckSubmissionRights() changes the DB sometimes)
+		}
+
+		validated, keyObtained = saveGradingResultsIntoDB(store, user, &requestData)
+		return nil
+	})
+	if apiError != service.NoError {
+		return apiError
+	}
+	service.MustNotBeError(err)
+
+	if validated && requestData.TaskToken.Converted.AccessSolutions != nil && !(*requestData.TaskToken.Converted.AccessSolutions) {
+		requestData.TaskToken.AccessSolutions = formdata.AnythingFromString(`"1"`)
+	}
+	requestData.TaskToken.PlatformName = srv.TokenConfig.PlatformName
+	newTaskToken, err := requestData.TaskToken.Sign(srv.TokenConfig.PrivateKey)
+	service.MustNotBeError(err)
+
+	service.MustNotBeError(render.Render(w, r, service.CreationSuccess(map[string]interface{}{
+		"task_token":   newTaskToken,
+		"bValidated":   validated,
+		"bKeyObtained": keyObtained,
+	})))
+	return service.NoError
+}
+
+func saveGradingResultsIntoDB(store *database.DataStore, user *database.User,
+	requestData *SaveGradeRequest) (validated, keyObtained bool) {
+	const todo = "todo"
+	score := requestData.ScoreToken.Converted.Score
+	userAnswerID := requestData.ScoreToken.Converted.UserAnswerID
+
+	// TODO: handle validation in a proper way (what did he mean??)
```

Comment:
I'm not sure because items.iValidationMin is int(11) while users_items/groups_attempts.iScore is float. Anyway, do we want to implement that here and now?